aptos-move/aptos-vm/src/tests/test_resolver_with_aggregator_liftings.rs

```diff
@@ -0,0 +1,366 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Error;
+use aptos_table_natives::{TableHandle, TableResolver};
+use aptos_types::{access_path::AccessPath, state_store::state_key::StateKey};
+use move_core_types::{
+    account_address::AccountAddress,
+    language_storage::StructTag,
+    metadata::Metadata,
+    resolver::ResourceResolver,
+    value::{LayoutTag, MoveStructLayout, MoveTypeLayout},
+};
+use move_vm_types::{
+    value_exchange::{
+        deserialize_and_exchange, AsIdentifier, ExchangeResult, IdentifierBuilder, ValueExchange,
+    },
+    values::{Struct, Value},
+};
+use std::{cell::RefCell, collections::BTreeMap, str::FromStr};
+
+/// Models a storage backend which only stores bytes of data. Storage can be
+/// resolved like anything else, to get a resource, but it does not do anything
+/// special.
+#[derive(Debug, Default)]
+struct Storage {
+    inner: BTreeMap<StateKey, Vec<u8>>,
+}
+
+impl ResourceResolver for Storage {
+    fn get_resource_with_metadata_and_layout(
+        &self,
+        address: &AccountAddress,
+        typ: &StructTag,
+        _metadata: &[Metadata],
+        _layout: Option<&MoveTypeLayout>,
+    ) -> anyhow::Result<(Option<Vec<u8>>, usize), anyhow::Error> {
+        let ap = AccessPath::resource_access_path(*address, typ.clone())
+            .expect("Access path for resource have to be valid");
+        let state_key = StateKey::access_path(ap);
+        Ok(match self.inner.get(&state_key) {
+            Some(blob) => (Some(blob.clone()), blob.len()),
+            None => (None, 0),
+        })
+    }
+}
+
+impl TableResolver for Storage {
+    fn resolve_table_entry_with_layout(
+        &self,
+        handle: &TableHandle,
+        key: &[u8],
+        _layout: Option<&MoveTypeLayout>,
+    ) -> Result<Option<Vec<u8>>, Error> {
+        let state_key = StateKey::table_item((*handle).into(), key.to_vec());
+        Ok(self.inner.get(&state_key).cloned())
+    }
+}
+
+/// Models a cache on top of storage. In addition to cached values, it also
+/// stores a lifting map of extracted aggregator values.
+#[derive(Debug, Default)]
+struct Cache {
+    liftings: RefCell<BTreeMap<u64, Value>>,
```

Comment:
Could we put TestExchange here and share the code with Exchange tests?
I think we want access to liftings for the test here though? but it's a refcell, and we would then have rc<refcell to also pass to the TestExchange constructor, so maybe not worth it - up to you - but just mentioning the possibility.

Reply:
We actually do not need TestExchange at all in this case, can just test it using liftings from cache?

Reply:
oh, cool.